### PR TITLE
[Merged by Bors] - TO-3114 add storage layer

### PR DESCRIPTION
**Pull request by Robert-Steiner**
Ticket:
- [TO-3114]

[TO-3114]: https://xainag.atlassian.net/browse/TO-3114?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: discovery_engine_core/web-api/src/storage.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    pub(crate) async fn update(
+        &self,
+        id: &UserId,
+        state: &CoiSystemState,
+    ) -> Result<(), GenericError> {
+        let mut tx = self.pool.begin().await?;
+
+        let serialized_state = bincode::serialize(state)?;
```
Comment (janpetschexain) [inline discovery_engine_core/web-api/src/storage.rs:68]: :art: those two lines could be switched to reduce the transaction "span"
```suggestion
        let serialized_state = bincode::serialize(state)?;
        
        let mut tx = self.pool.begin().await?;
```

File: discovery_engine_core/web-api/src/migrations/20220811080009_user.sql [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+--
+--  This program is distributed in the hope that it will be useful,
+--  but WITHOUT ANY WARRANTY; without even the implied warranty of
+--  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+--  GNU Affero General Public License for more details.
+--
+--  You should have received a copy of the GNU Affero General Public License
+--  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+CREATE TABLE IF NOT EXISTS user_state (
```
[the remark this answers is not in the document]
Comment (Robert-Steiner) [reply discovery_engine_core/web-api/src/migrations/20220811080009_user.sql:15]: It doesn't seem set in stone, but the naming convention for identifiers seems to be lowercase. For example, the names in the Postgres documentation examples are always lowercase.

File: discovery_engine_core/web-api/src/storage.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    ) -> Result<(), GenericError> {
+        let serialized_state = bincode::serialize(state)?;
+
+        let mut tx = self.pool.begin().await?;
+
+        sqlx::query(
+            "INSERT INTO user_state(id, state) VALUES ($1, $2)
+                ON CONFLICT (id) DO UPDATE SET state = EXCLUDED.state;",
+        )
+        .bind(id.to_string())
```
Comment (janpetschexain) [inline discovery_engine_core/web-api/src/storage.rs:74]: :art: same
```suggestion
        .bind(id.as_ref())
```

**Review by janpetschexain — APPROVED** [review]

**Comment by Robert-Steiner** [comment]
bors merge

**Pull request closed without merging**